src/EventStore.Core/Services/Archiver/Storage/FileSystemArchiveStorage.cs

```diff
@@ -16,11 +21,78 @@ public FileSystemArchiveStorage(FileSystemOptions options) {
 		_archivePath = options.Path;
 	}
 
-	public ValueTask StoreChunk(string pathToSourceChunk) {
-		var fileName = Path.GetFileName(pathToSourceChunk);
-		var pathToDestnationChunk = Path.Combine(_archivePath, fileName);
-		Log.Information("Copying file {Path} to {Destination}", pathToSourceChunk, pathToDestnationChunk);
-		File.Copy(pathToSourceChunk, pathToDestnationChunk);
-		return ValueTask.CompletedTask;
+	public async ValueTask<bool> StoreChunk(string chunkPath, CancellationToken ct) {
+		try {
+			var destinationPath = Path.Combine(_archivePath, Path.GetFileName(chunkPath));
+			var tempPath = $"{destinationPath}.tmp";
+
+			if (File.Exists(destinationPath))
+				File.Delete(destinationPath);
+
+			if (File.Exists(tempPath))
+				File.Delete(tempPath);
+
+			await using var source = File.Open(
+				path: chunkPath,
+				options: new FileStreamOptions {
+					Mode = FileMode.Open,
+					Access = FileAccess.Read,
+					Share = FileShare.Read,
+					Options = FileOptions.SequentialScan | FileOptions.Asynchronous
+				});
+
+			await using var destination = File.Open(
+				path: tempPath,
+				options: new FileStreamOptions {
+					Mode = FileMode.CreateNew,
+					Access = FileAccess.ReadWrite,
+					Share = FileShare.None,
+					Options = FileOptions.Asynchronous,
+					PreallocationSize = new FileInfo(chunkPath).Length
+				});
+
+			await source.CopyToAsync(destination, ct);
+
+			File.Move(tempPath, destinationPath);
+
+			return true;
+		} catch (OperationCanceledException) {
+			throw;
+		} catch (Exception ex) {
+			if (!File.Exists(chunkPath))
+				throw new ChunkDeletedException();
+
+			Log.Error(ex, "Error while storing chunk: {chunkFile}", Path.GetFileName(chunkPath));
+			return false;
+		}
+	}
+
+	public ValueTask<bool> RemoveChunks(int chunkStartNumber, int chunkEndNumber, string exceptChunk, CancellationToken ct) {
+		try {
+			var directoryInfo = new DirectoryInfo(_archivePath);
+			for (var chunkNumber = chunkStartNumber; chunkNumber <= chunkEndNumber; chunkNumber++) {
+				foreach (var file in directoryInfo.EnumerateFiles($"chunk-{chunkNumber:000000}.*")) {
+					if (file.Name == exceptChunk)
+						continue;
+
+					File.Delete(file.FullName);
+				}
+			}
+		} catch (OperationCanceledException) {
+			throw;
+		} catch (Exception ex) {
+			Log.Error(ex, "Error while removing chunks in range: {chunkStartNumber}-{chunkEndNumber} (except {chunk})",
+				chunkStartNumber, chunkEndNumber, exceptChunk);
+			return ValueTask.FromResult(false);
+		}
+
+		return ValueTask.FromResult(true);
+	}
+
+	public IAsyncEnumerable<string> ListChunks(CancellationToken ct) {
+		return new DirectoryInfo(_archivePath)
+			.EnumerateFiles("chunk-*")
```

Comment:
not necessarily now but perhaps we should at least add a note that a `IVersionedFileNamingStrategy` should probably drive the filtering here (and elsewhere in this file where the "chunk" prefix/pattern is hard coded

Reply:
good idea, fixed (but not tested yet)